IdentityCore/src/logger/MSIDLogger.m

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDLogger.h"
+#import "MSIDLogger+Internal.h"
+#import "MSIDVersion.h"
+#import "MSIDDeviceId.h"
+
+@interface MSIDLogger()
+{
+    MSIDLogCallback _callback;
+}
+
+@end
+
+@implementation MSIDLogger
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    // The default log level should be info, anything more restrictive then this
+    // and we'll probably not have enough diagnostic information, however verbose
+    // will most likely be too noisy for most usage.
+    self.level = MSIDLogLevelInfo;
+    self.PiiLoggingEnabled = NO;
+    
+    return self;
+}
+
++ (MSIDLogger *)sharedLogger
+{
+    static dispatch_once_t once;
+    static MSIDLogger *s_logger;
+    
+    dispatch_once(&once, ^{
+        s_logger = [MSIDLogger new];
+    });
+    
+    return s_logger;
+}
+
+- (void)setCallback:(MSIDLogCallback)callback
+{
+    static dispatch_once_t once;
+    
+    if (self->_callback != nil)
+    {
+        @throw @"MSID logging callback can only be set once per process and should never changed once set.";
+    }
+    
+    dispatch_once(&once, ^{
+        self->_callback = callback;
+    });
+}
+
+
+@end
+
+@implementation MSIDLogger (Internal)
+
+static NSDateFormatter *s_dateFormatter = nil;
+
++ (void)initialize
+{
+    s_dateFormatter = [[NSDateFormatter alloc] init];
+    [s_dateFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];
+    [s_dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
+}
+
+- (void)logLevel:(MSIDLogLevel)level
+         context:(id<MSIDRequestContext>)context
+   correlationId:(NSUUID *)correlationId
+           isPII:(BOOL)isPii
+          format:(NSString *)format, ...
```

Comment:
instead of context here, why not just log component here?

Reply:
Discussed offline that we don't want to pass log component and correlation id separately